### PR TITLE
[gosrc2cpg] - Global variable as member nodes of package level TypeDecl

### PR DESCRIPTION
**Pull request by pandurangpatil**
Follow up PR to the PR #3734 to model global variables as member nodes of Package level TypeDecl, and its access is converted to `field access` call.

### CODE REVIEW

File: joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/astcreation/AstForGenDeclarationCreator.scala [hunk 1]
```diff
@@ -75,23 +91,43 @@ trait AstForGenDeclarationCreator(implicit withSchemaValidation: ValidationMode)
...
-      rhsParserNode,
-      code,
-      Operators.assignment,
-      Operators.assignment,
-      DispatchTypes.STATIC_DISPATCH,
-      None,
-      Some(rhsTypeFullName)
+    if (globalStatements) {
+      astForGlobalVarAndConstants(rhsTypeFullName, lhsParserNode, Some(rhsAst))
+      (Ast(), Ast())
```
Comment (DavidBakerEffendi) [inline joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/astcreation/AstForGenDeclarationCreator.scala:101]: Perhaps here would be a good place to have a field access-based assignment to the respective `FieldIdentifier`?

**Review by DavidBakerEffendi — CHANGES_REQUESTED** [review]
One thing I noticed when trying to bring Python to a similar design, was that the data-flow engine sees field accesses differently when looking at slices. 

While GoLang is different, and does not have expressions on a module-level necessarily, it may be good to look at the PR to see how I dealt with this pattern: https://github.com/joernio/joern/pull/3750

File: joern-cli/frontends/gosrc2cpg/src/test/scala/io/joern/go2cpg/passes/ast/VariableReferencingTests.scala [hunk 2]
```diff
@@ -43,26 +43,21 @@ class VariableReferencingTests extends GoCodeToCpgSuite {
...
-        |y:=x
+        | y := x
         |}
         |""".stripMargin)
 
-    "test local variable exists" in {
-      val List(localNode) = cpg.method("main.Test0.go").local.nameExact("x").l
-      localNode.code shouldBe "x"
-      localNode.closureBindingId shouldBe None
+    "check Global Member node" in {
```
[the remark this answers is not in the document]
Comment (pandurangpatil) [reply joern-cli/frontends/gosrc2cpg/src/test/scala/io/joern/go2cpg/passes/ast/VariableReferencingTests.scala:50]: there are a couple of other places I have already added the unit tests for `Operators.fieldAccess` largely in the previous PR. In this PR I have made handling for the global variable access from the same package as that was being done without the package name alias.

https://github.com/joernio/joern/pull/3756/files#diff-fdbf251d852135d11159b2009ec3e07361c347a02ad8aec38fe5356e5c16b814R39

File: joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/astcreation/AstForGenDeclarationCreator.scala (hunk 1 above)
Comment (pandurangpatil) [reply joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/astcreation/AstForGenDeclarationCreator.scala:101]: I didn't get the question.

Comment (pandurangpatil) [reply joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/astcreation/AstForGenDeclarationCreator.scala:101]: If you are pointing at the modelling of the values assigned to the global variable. For that, I need to model it as a constructor method node and initialise the global member variables as part of the body. 

I will handle that part in separate PR along with respective unit tests for the same. I should have mentioned this in the TODO.

File: joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/astcreation/AstCreator.scala [hunk 3]
```diff
@@ -63,15 +63,22 @@ class AstCreator(val relPathFileName: String, val parserResult: ParserResult, go
     scope.pushNewScope(fakeGlobalMethodForFile)
     val blockNode_   = blockNode(rootNode, Defines.empty, Defines.anyTypeName)
     val methodReturn = methodReturnNode(rootNode, Defines.anyTypeName)
-    val declsAsts    = rootNode.json(ParserKeys.Decls).arr.flatMap(item => astForNode(item)).toList
+    val declsAsts = rootNode
+      .json(ParserKeys.Decls)
+      .arr
+      .flatMap(item => {
```
[the remark this answers is not in the document]
Comment (pandurangpatil) [reply joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/astcreation/AstCreator.scala]: done

File: joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/astcreation/AstForTypeDeclCreator.scala [hunk 4]
```diff
@@ -79,14 +79,21 @@ trait AstForTypeDeclCreator(implicit withSchemaValidation: ValidationMode) { thi
...
-    val fieldIdAst = Ast(fieldIdentifierNode)
     val callNode =
       newOperatorCallNode(Operators.fieldAccess, info.code, Some(fieldTypeFullName), line(info), column(info))
-    Seq(callAst(callNode, identifierAsts ++ Seq(fieldIdAst)))
+    Seq(
+      callAst(
+        callNode,
+        identifierAsts ++ Seq(
+          Ast(
+            NewFieldIdentifier()
```
[the remark this answers is not in the document]
Comment (pandurangpatil) [reply joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/astcreation/AstForTypeDeclCreator.scala]: done

**Review by DavidBakerEffendi — APPROVED** [review]

**Pull request merged**